File: app/src/main/java/com/github/gotify/messages/provider/MessageImageCombiner.java
package com.github.gotify.messages.provider;

import com.github.gotify.client.model.Application;
import com.github.gotify.client.model.Message;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class MessageImageCombiner {

    List<MessageWithImage> combine(List<Message> messages, List<Application> applications) {
        Map<Long, String> appIdToImage = appIdToImage(applications);

        List<MessageWithImage> result = new ArrayList<>();

        for (Message message : messages) {
            MessageWithImage messageWithImage = new MessageWithImage();

            messageWithImage.message = message;
            messageWithImage.image = appIdToImage.get(message.getAppid());

            result.add(messageWithImage);
        }

        return result;
    }

    public static Map<Long, String> appIdToImage(List<Application> applications) {
        Map<Long, String> map = new ConcurrentHashMap<>();
        for (Application app : applications) {
            map.put(app.getId(), app.getImage());
        }
        return map;
    }
}

File: app/src/main/kotlin/com/github/gotify/messages/provider/MessageImageCombiner.kt
package com.github.gotify.messages.provider

import com.github.gotify.client.model.Application
import com.github.gotify.client.model.Message

internal object MessageImageCombiner {
    fun combine(messages: List<Message>, applications: List<Application>): List<MessageWithImage> {
        val appIdToImage = appIdToImage(applications)
        return messages.map { MessageWithImage(message = it, image = appIdToImage[it.appid]) }
    }

    private fun appIdToImage(applications: List<Application>): Map<Long, String> {
        val map = mutableMapOf<Long, String>()
        applications.forEach {
            map[it.id] = it.image
        }
        return map
    }
}
